server/src/main/java/org/elasticsearch/index/seqno/RetentionLeaseSyncAction.java

```diff
@@ -0,0 +1,206 @@
+        return LOGGER;
+    }
+
+    @Inject
+    public RetentionLeaseSyncAction(
+            final Settings settings,
+            final TransportService transportService,
+            final ClusterService clusterService,
+            final IndicesService indicesService,
+            final ThreadPool threadPool,
+            final ShardStateAction shardStateAction,
+            final ActionFilters actionFilters,
+            final IndexNameExpressionResolver indexNameExpressionResolver) {
+        super(
+                settings,
+                ACTION_NAME,
+                transportService,
+                clusterService,
+                indicesService,
+                threadPool,
+                shardStateAction,
+                actionFilters,
+                indexNameExpressionResolver,
+                RetentionLeaseSyncAction.Request::new,
+                RetentionLeaseSyncAction.Request::new,
+                ThreadPool.Names.MANAGEMENT);
+    }
+
+    /**
+     * Sync the specified retention leases for the specified shard. The callback is invoked when the sync succeeds or fails.
+     *
+     * @param shardId         the shard to sync
+     * @param retentionLeases the retention leases to sync
+     * @param listener        the callback to invoke when the sync completes normally or abnormally
+     */
+    public void syncRetentionLeasesForShard(
+            final ShardId shardId,
+            final Collection<RetentionLease> retentionLeases,
+            final ActionListener<ReplicationResponse> listener) {
+        Objects.requireNonNull(shardId);
+        Objects.requireNonNull(retentionLeases);
+        Objects.requireNonNull(listener);
+        final ThreadContext threadContext = threadPool.getThreadContext();
+        try (ThreadContext.StoredContext ignore = threadContext.stashContext()) {
+            // we have to execute under the system context so that if security is enabled the sync is authorized
+            threadContext.markAsSystemContext();
+            execute(
+                    new RetentionLeaseSyncAction.Request(shardId, retentionLeases),
+                    ActionListener.wrap(
+                            listener::onResponse,
+                            e -> {
+                                if (ExceptionsHelper.unwrap(e, AlreadyClosedException.class, IndexShardClosedException.class) == null) {
+                                    getLogger().warn(new ParameterizedMessage("{} retention lease sync failed", shardId), e);
+                                }
+                                listener.onFailure(e);
+                            }));
+        }
+    }
+
+    @Override
+    protected WritePrimaryResult<Request, Response> shardOperationOnPrimary(final Request request, final IndexShard primary) {
+        Objects.requireNonNull(request);
+        Objects.requireNonNull(primary);
+        // we flush to ensure that retention leases are committed
+        flush(primary);
```

Comment:
If the primary recovers from the safe commit, we will lose the committed retention leases. Should we copy them in `Store#trimUnsafeCommits` ?

Reply:
Good call. Let us get that in a follow-up.